samples/performance/constant_data/constant_data_tutorial.md

```diff
@@ -0,0 +1,399 @@
+
+In theory, this data can be anything we want it to be, for instance it can be used for things such as calculating where an object should be placed inside our world, or computing the overall brightness of an object based on the lights in the scene.
+
+It differs from other data (e.g. input vertex data) in the sense that it remains *constant* across every shader invocation of a draw call.
+
+This is important as because of this assumption, the data can be **shared** between shader stages, as we know it isn't going to be changed throughout the runtime of a single draw call in a render pipeline.
+
+The next section aims to cover the constant data theory, starting at the shader level before moving to the basics of how to plug in your data using Vulkan.
+
+### **Constant data in Vulkan shaders**
+
+Constant data is implemented in our shader code by using global variables.
+
+**Global variables** have the following format:
+`<layout> <storage> <type> <variable_name>`.  
+
+Take this vertex shader for example:
+
+```
+layout(location = 0) in vec4 position;
+
+layout(set = 0, binding = 0) uniform ConstantData
+{
+    mat4 model;
+} constant_data;
+
+layout(location = 0) out vec4 o_pos;
+```
+
+We can see three global variables, each with a different **storage type**:
+* Inputs (`in`)
+* Uniforms (`uniform`)
+* Outputs (`out`)
+
+#### *Varying Types*
+
+The global variables that use inputs (`in`) and outputs (`out`) are values that *may vary* from one shader invocation to the next, therefore they **shouldn't** be used for constant data. They require a `layout location` which is used to identify a particular input/output.
+
+They have slightly different rules for what they do depending on the shader stage, and have slightly different restrictions on the types of data it can represent. However, generally their use is to feed values from one stage to the next (e.g. from vertex shader to fragment shader). 
+
+*You can read more about shader stage inputs and outputs [here](https://www.khronos.org/opengl/wiki/Type_Qualifier_(GLSL)#Shader_stage_inputs_and_outputs).*
+
+#### *Constant Types*
+
+Constant types are global variables that have either the `uniform` or `buffer` storage type, these are *uniform buffer objects* and *shader storage buffer objects* respectively. They describe data which remains constant across an entire draw call, meaning that the values stay the same across the different shader stages and shader invocations.
+
+**Uniform buffer objects (UBOs)** are the more commonly used of the two. They can hold up to 16KB of data. It is important to note that they are not *compile-time* constant, so they are not given the `const` type. However, trying to edit the data in shader code will result in a compile-time error. These values use a `layout binding` and, when working with multiple `VkDescriptorSet`s, we will even give it a `layout set`. *You can read more about UBOs [here](https://www.khronos.org/opengl/wiki/Uniform_Buffer_Object).*
```

Comment:
> They can hold up to 16KB of data.

On Arm maybe. Our limit is much higher, as are other vendors.

Reply:
This is written against the GLSL spec, these are the guaranteed limits. I believe most implementations of Vulkan let you use more. Do I change the wording? It probably isn't a great idea to maintain a list of these limits